sdl-debezium/src/main/scala/io/smartdatalake/workflow/dataobject/DebeziumCdcDataObject.scala

```diff
@@ -0,0 +1,329 @@
+
+      completionCallback.error.foreach(err => throw new Exception(err))
+
+      changeConsumer.records
+
+    }
+
+    def createEmptyDataFrame(): DataFrame = {
+
+        val schemaProperties = debeziumPropertiesForEngine
+
+        Seq("offset.storage", "offset.storage.sdlb.data.object.id").foreach(schemaProperties.remove(_))
+
+        schemaProperties.put("offset.storage", "org.apache.kafka.connect.storage.MemoryOffsetBackingStore")
+
+        val records = getRecordsFromDebeziumEngine(schemaProperties, changeConsumer = new DebeziumSchemaConsumer)
+
+        val df = DebeziumEventConverter.convert(records)(spark)
+
+        val schema = df.schema
+
+        spark.createDataFrame(new util.ArrayList[Row](), schema)
+
+    }
+
+    if (context.isExecPhase) {
+
+      val records = getRecordsFromDebeziumEngine(debeziumPropertiesForEngine, changeConsumer = new DebeziumChangeConsumer)
+
+      records.headOption match {
+        case Some(_) => {
+          DebeziumEventConverter.convert(records)(spark)
+        }
+        case None => createEmptyDataFrame()
+      }
+
+    } else {
+      createEmptyDataFrame()
+    }
+
+  }
+
+  private[smartdatalake] var incrementalState: mutable.Map[String, String] = mutable.Map()
+
+  /**
+   * To implement incremental processing this function is called to initialize the DataObject with its state from the last increment.
+   * The state is just a string. It's semantics is internal to the DataObject.
+   * Note that this method is called on initializiation of the SmartDataLakeBuilder job (init Phase) and for streaming execution after every execution of an Action involving this DataObject (postExec).
+   *
+   * @param state Internal state of last increment. If None then the first increment (may be a full increment) is delivered.
+   */
+  override def setState(state: Option[String])(implicit context: ActionPipelineContext): Unit = {
+
+    state match {
+      case Some(s) =>
+        s.split(",").foreach { pair =>
+          val Array(key, value) = pair.split(":")
+          incrementalState.put(key, value)
+        }
+    }
+  }
+
+  /**
+   * Return the state of the last increment or empty if no increment was processed.
+   */
+  override def getState: Option[String] = {
+    val state = incrementalState.map { case (key, value) =>
+      s"$key:$value"
+    }.mkString(",")
+
+    Some(state)
+  }
+}
+
+object DebeziumCdcDataObject extends FromConfigFactory[DataObject] {
+  override def fromConfig(config: Config)(implicit instanceRegistry: InstanceRegistry): DebeziumCdcDataObject = {
+    extract[DebeziumCdcDataObject](config)
+  }
+}
+
+/**
+ * Helper object to convert from debezium events in SourceRecord format to sdlb compatible spark dataframe
+ */
+private object DebeziumEventConverter {
+
+  def convert(records: Seq[SourceRecord])(implicit spark: SparkSession): DataFrame = {
+
+    val sparkSchema = inferSparkSchema(records.head.valueSchema())
+    val rows = records.map {
```

Comment:
records.map(recordToRow)

would also be possible and more concise, just so you know